internal/fnruntime/runner.go

```diff
@@ -201,36 +201,38 @@ type FunctionRunner struct {
 
 func (fr *FunctionRunner) Filter(input []*yaml.RNode) (output []*yaml.RNode, err error) {
 	pr := printer.FromContextOrDie(fr.ctx)
+	var fnOutput strings.Builder
 	if !fr.disableCLIOutput {
 		if fr.wasm {
-			pr.Printf("[RUNNING] WASM %q", fr.name)
+			fnOutput.WriteString(fmt.Sprintf("[RUNNING] WASM %q", fr.name))
 		} else {
-			pr.Printf("[RUNNING] %q", fr.name)
+			fnOutput.WriteString(fmt.Sprintf("[RUNNING] %q", fr.name))
 		}
 		if fr.displayResourceCount {
-			pr.Printf(" on %d resource(s)", len(input))
+			fnOutput.WriteString(fmt.Sprintf(" on %d resource(s)", len(input)))
 		}
-		pr.Printf("\n")
+		fnOutput.WriteString("\n")
```

Comment:
If we don't write the `fnOutput` to the printer here, the end users in CLI case won't see the `[RUNNING]....` message before running the function. Instead, they will see it in the end. Seeing it before a function is actually invoked is important from the UX perspective especially in case of running a function for the first time (because it may take long time due to downloading the function image etc).

General rule: any diagnostic info about running something should be printed as real time as possible.

Reply:
Another thing I realized: This is important for a `kpt CLI` user because they see the feedback in realtime and may not be important for `porch` because a porch user sees the output at the end.

May be passing a simple printer (that uses buffered output) for porch invocation might work better leaving the current code untouched ?

Reply:
> May be passing a simple printer (that uses buffered output) for porch invocation might work better leaving the current code untouched ?

I think this makes sense. I will look further into this.